<a href="https://colab.research.google.com/github/22f1000003/22f1000003/blob/main/W2_L3demonstration_of_sklearn.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

the objective of colab is to demonstrate sklearn dataset API
Recall that it has three APIs


1.   Loaders (load_*) load small standard datasets bundied with sklearn.
2. Fetchers ( fetch_*) fetch large datasets from the internet and loads them in memory
3. Generater (generate_*) generate controlled synthetic dataset



**Loaders**

Loading iris dataset

In [ ]:
from sklearn.datasets import load_iris
data = load_iris()

Tjis return a bunch Object data which is dictionary like object with following attributes
1. data , which has the feature matrix
2. target ,  which is the lable vector
3. feature_names contain name of the features
4. target_names Contain the names of the classes.
5. DESCR has the full description of dataset

In [ ]:
type(data)

sklearn.utils._bunch.Bunch

We can acess them one by one and examine their contents. For exaple, we can access feature_names as follows

In [ ]:
data.feature_names

['sepal length (cm)',
 'sepal width (cm)',
 'petal length (cm)',
 'petal width (cm)']

we can see the name of the features in this dataset

Let's examine the names of the lables

In [ ]:
data.target_names

array(['setosa', 'versicolor', 'virginica'], dtype='<U10')

There are three classes setosa , wersicolor , virginica

the feature matrix can be accessed as follows data.data Let's look at the first five examples in feature matrix

In [ ]:
data.data[:5]

array([[5.1, 3.5, 1.4, 0.2],
       [4.9, 3. , 1.4, 0.2],
       [4.7, 3.2, 1.3, 0.2],
       [4.6, 3.1, 1.5, 0.2],
       [5. , 3.6, 1.4, 0.2]])

we can observe 4 feature per example
Let's examine the shape of the feature matrix

In [ ]:
data.data.shape

(150, 4)

THere are 150 example and each example has 4 features.
finally we will examine the label vector and its shape

In [ ]:
data.target

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

There are 50 exaples each from three classes; 0, 1, and 2.
we can read additional documentation about loads_iris in the following manner.

In [ ]:
?load_iris

in this way we can load and examine different datasets.
we can obtain feature matrix and label or target from load_iris and other ;oaders in general by setting return_x_y argument to true

In [ ]:
feature_matrix , label_vector = load_iris(return_X_y = True)
print('shape of feature matrix', feature_matrix.shape)
print('shape of label vector' , label_vector.shape)

shape of feature matrix (150, 4)
shape of label vector (150,)


**Loading diabetes dataset

from sklearn.datasets import load_diabetes
data = load_diabetes()

In [ ]:
from sklearn.datasets import load_diabetes
data = load_diabetes()

additional details about the loader can be accessed from the documentation

In [ ]:
?load_diabetes

**load_diabetes**
1. step Load the dataset and obtain a bunch object.

In [ ]:
type(data)

sklearn.utils._bunch.Bunch

Examine the bunch object
Look at the Description of the dataset

In [ ]:
data.DESCR

'.. _diabetes_dataset:\n\nDiabetes dataset\n----------------\n\nTen baseline variables, age, sex, body mass index, average blood\npressure, and six blood serum measurements were obtained for each of n =\n442 diabetes patients, as well as the response of interest, a\nquantitative measure of disease progression one year after baseline.\n\n**Data Set Characteristics:**\n\n:Number of Instances: 442\n\n:Number of Attributes: First 10 columns are numeric predictive values\n\n:Target: Column 11 is a quantitative measure of disease progression one year after baseline\n\n:Attribute Information:\n    - age     age in years\n    - sex\n    - bmi     body mass index\n    - bp      average blood pressure\n    - s1      tc, total serum cholesterol\n    - s2      ldl, low-density lipoproteins\n    - s3      hdl, high-density lipoproteins\n    - s4      tch, total cholesterol / HDL\n    - s5      ltg, possibly log of serum triglycerides level\n    - s6      glu, blood sugar level\n\nNote: Each of thes

In [ ]:
feature_matrix , label_vector = load_diabetes(return_X_y=True )

In [ ]:
feature_matrix.shape


(442, 10)

In [ ]:
label_vector.shape

(442,)

**find the five example from the feature matrix**

In [ ]:
data.data[:5]

array([[ 0.03807591,  0.05068012,  0.06169621,  0.02187239, -0.0442235 ,
        -0.03482076, -0.04340085, -0.00259226,  0.01990749, -0.01764613],
       [-0.00188202, -0.04464164, -0.05147406, -0.02632753, -0.00844872,
        -0.01916334,  0.07441156, -0.03949338, -0.06833155, -0.09220405],
       [ 0.08529891,  0.05068012,  0.04445121, -0.00567042, -0.04559945,
        -0.03419447, -0.03235593, -0.00259226,  0.00286131, -0.02593034],
       [-0.08906294, -0.04464164, -0.01159501, -0.03665608,  0.01219057,
         0.02499059, -0.03603757,  0.03430886,  0.02268774, -0.00936191],
       [ 0.00538306, -0.04464164, -0.03638469,  0.02187239,  0.00393485,
         0.01559614,  0.00814208, -0.00259226, -0.03198764, -0.04664087]])

find out the lable of the first five examples

In [ ]:
data.target[:5]

array([151.,  75., 141., 206., 135.])

find out name of class lable

In [ ]:
data.feature_names


['age', 'sex', 'bmi', 'bp', 's1', 's2', 's3', 's4', 's5', 's6']

**Loading digits dataset**

from sklearn.datasets import load_digits
data =load_digits


In [ ]:
from sklearn.datasets import load_digits
data = load_digits()

In [ ]:
?load_digits

**Look at the description of the dataset**

In [ ]:
data.DESCR

".. _digits_dataset:\n\nOptical recognition of handwritten digits dataset\n--------------------------------------------------\n\n**Data Set Characteristics:**\n\n:Number of Instances: 1797\n:Number of Attributes: 64\n:Attribute Information: 8x8 image of integer pixels in the range 0..16.\n:Missing Attribute Values: None\n:Creator: E. Alpaydin (alpaydin '@' boun.edu.tr)\n:Date: July; 1998\n\nThis is a copy of the test set of the UCI ML hand-written digits datasets\nhttps://archive.ics.uci.edu/ml/datasets/Optical+Recognition+of+Handwritten+Digits\n\nThe data set contains images of hand-written digits: 10 classes where\neach class refers to a digit.\n\nPreprocessing programs made available by NIST were used to extract\nnormalized bitmaps of handwritten digits from a preprinted form. From a\ntotal of 43 people, 30 contributed to the training set and different 13\nto the test set. 32x32 bitmaps are divided into nonoverlapping blocks of\n4x4 and the number of on pixels are counted in each bl

In [ ]:
print(type(data))

<class 'sklearn.utils._bunch.Bunch'>


In [ ]:
data.data[:1]

array([[ 0.,  0.,  5., 13.,  9.,  1.,  0.,  0.,  0.,  0., 13., 15., 10.,
        15.,  5.,  0.,  0.,  3., 15.,  2.,  0., 11.,  8.,  0.,  0.,  4.,
        12.,  0.,  0.,  8.,  8.,  0.,  0.,  5.,  8.,  0.,  0.,  9.,  8.,
         0.,  0.,  4., 11.,  0.,  1., 12.,  7.,  0.,  0.,  2., 14.,  5.,
        10., 12.,  0.,  0.,  0.,  0.,  6., 13., 10.,  0.,  0.,  0.]])

**Find out shape of feature matrix

In [ ]:
feature_matrix , label_vector = load_digits(return_X_y = True)

In [ ]:
feature_matrix.shape


(1797, 64)

In [ ]:
label_vector.shape

(1797,)

**Look at the first five example from the feature matrix**

In [ ]:
feature_matrix[:1]

array([[ 0.,  0.,  5., 13.,  9.,  1.,  0.,  0.,  0.,  0., 13., 15., 10.,
        15.,  5.,  0.,  0.,  3., 15.,  2.,  0., 11.,  8.,  0.,  0.,  4.,
        12.,  0.,  0.,  8.,  8.,  0.,  0.,  5.,  8.,  0.,  0.,  9.,  8.,
         0.,  0.,  4., 11.,  0.,  1., 12.,  7.,  0.,  0.,  2., 14.,  5.,
        10., 12.,  0.,  0.,  0.,  0.,  6., 13., 10.,  0.,  0.,  0.]])

**find out the name of feature names**

In [ ]:
data.feature_names
print(data.feature_names)

['pixel_0_0', 'pixel_0_1', 'pixel_0_2', 'pixel_0_3', 'pixel_0_4', 'pixel_0_5', 'pixel_0_6', 'pixel_0_7', 'pixel_1_0', 'pixel_1_1', 'pixel_1_2', 'pixel_1_3', 'pixel_1_4', 'pixel_1_5', 'pixel_1_6', 'pixel_1_7', 'pixel_2_0', 'pixel_2_1', 'pixel_2_2', 'pixel_2_3', 'pixel_2_4', 'pixel_2_5', 'pixel_2_6', 'pixel_2_7', 'pixel_3_0', 'pixel_3_1', 'pixel_3_2', 'pixel_3_3', 'pixel_3_4', 'pixel_3_5', 'pixel_3_6', 'pixel_3_7', 'pixel_4_0', 'pixel_4_1', 'pixel_4_2', 'pixel_4_3', 'pixel_4_4', 'pixel_4_5', 'pixel_4_6', 'pixel_4_7', 'pixel_5_0', 'pixel_5_1', 'pixel_5_2', 'pixel_5_3', 'pixel_5_4', 'pixel_5_5', 'pixel_5_6', 'pixel_5_7', 'pixel_6_0', 'pixel_6_1', 'pixel_6_2', 'pixel_6_3', 'pixel_6_4', 'pixel_6_5', 'pixel_6_6', 'pixel_6_7', 'pixel_7_0', 'pixel_7_1', 'pixel_7_2', 'pixel_7_3', 'pixel_7_4', 'pixel_7_5', 'pixel_7_6', 'pixel_7_7']


**find out name of class labels**

In [ ]:
print(data.target_names)

[0 1 2 3 4 5 6 7 8 9]


**Exercise**

**Experiment with other dataset.loaders e.g. load_wise, load_breast_cancer and load_linnerud**

**load_wine**

In [ ]:
from sklearn.datasets import load_wine
data = load_wine()

In [ ]:
type(data)

sklearn.utils._bunch.Bunch

In [ ]:
feature_matrix , label_vector = load_wine(return_X_y=True)

In [ ]:
feature_matrix[:1]

array([[1.423e+01, 1.710e+00, 2.430e+00, 1.560e+01, 1.270e+02, 2.800e+00,
        3.060e+00, 2.800e-01, 2.290e+00, 5.640e+00, 1.040e+00, 3.920e+00,
        1.065e+03]])

In [ ]:
feature_matrix.shape

(178, 13)

In [ ]:
label_vector.shape

(178,)

In [ ]:
len(data.target)

178

In [ ]:
data.target

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2])

In [ ]:
data.DESCR

'.. _wine_dataset:\n\nWine recognition dataset\n------------------------\n\n**Data Set Characteristics:**\n\n:Number of Instances: 178\n:Number of Attributes: 13 numeric, predictive attributes and the class\n:Attribute Information:\n    - Alcohol\n    - Malic acid\n    - Ash\n    - Alcalinity of ash\n    - Magnesium\n    - Total phenols\n    - Flavanoids\n    - Nonflavanoid phenols\n    - Proanthocyanins\n    - Color intensity\n    - Hue\n    - OD280/OD315 of diluted wines\n    - Proline\n    - class:\n        - class_0\n        - class_1\n        - class_2\n\n:Summary Statistics:\n\n============================= ==== ===== ======= =====\n                                Min   Max   Mean     SD\n============================= ==== ===== ======= =====\nAlcohol:                      11.0  14.8    13.0   0.8\nMalic Acid:                   0.74  5.80    2.34  1.12\nAsh:                          1.36  3.23    2.36  0.27\nAlcalinity of Ash:            10.6  30.0    19.5   3.3\nMagnesium:      

In [ ]:
data.target_names

array(['class_0', 'class_1', 'class_2'], dtype='<U7')

In [ ]:
print(data.data[:5])

[[1.423e+01 1.710e+00 2.430e+00 1.560e+01 1.270e+02 2.800e+00 3.060e+00
  2.800e-01 2.290e+00 5.640e+00 1.040e+00 3.920e+00 1.065e+03]
 [1.320e+01 1.780e+00 2.140e+00 1.120e+01 1.000e+02 2.650e+00 2.760e+00
  2.600e-01 1.280e+00 4.380e+00 1.050e+00 3.400e+00 1.050e+03]
 [1.316e+01 2.360e+00 2.670e+00 1.860e+01 1.010e+02 2.800e+00 3.240e+00
  3.000e-01 2.810e+00 5.680e+00 1.030e+00 3.170e+00 1.185e+03]
 [1.437e+01 1.950e+00 2.500e+00 1.680e+01 1.130e+02 3.850e+00 3.490e+00
  2.400e-01 2.180e+00 7.800e+00 8.600e-01 3.450e+00 1.480e+03]
 [1.324e+01 2.590e+00 2.870e+00 2.100e+01 1.180e+02 2.800e+00 2.690e+00
  3.900e-01 1.820e+00 4.320e+00 1.040e+00 2.930e+00 7.350e+02]]


In [ ]:
print(data.target[:5])

[0 0 0 0 0]


In [ ]:
print(data.target[:67])

[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1]


**breast cancer**

In [ ]:
from sklearn.datasets import load_breast_cancer
data = load_breast_cancer()

In [ ]:
type(data)

sklearn.utils._bunch.Bunch

In [ ]:
feacher_matrix , label_vector = load_breast_cancer(return_X_y = True)

In [ ]:
feature_matrix.shape

(178, 13)

In [ ]:
label_vector.shape

(178,)

In [ ]:
data.target[:23]

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1,
       0])

**malignant have [0] label and benign [1]**

In [ ]:
data.target_names

array(['malignant', 'benign'], dtype='<U9')

**Fetchers**

**fetch_california_housing**
step 1: Import the library and access the documentation

In [ ]:
from sklearn.datasets import fetch_california_housing
?fetch_california_housing


**Note that the fetch_* also returns a Bunch object just like loaders.
we can examine various attributes of this dataset on the datasets in loaders**
1. load the dataset and obtain a Bunch object.


In [ ]:
housing_data = fetch_california_housing()

**Examine the bunch object**
1. look at the description of dataset

In [ ]:
housing_data.DESCR

'.. _california_housing_dataset:\n\nCalifornia Housing dataset\n--------------------------\n\n**Data Set Characteristics:**\n\n:Number of Instances: 20640\n\n:Number of Attributes: 8 numeric, predictive attributes and the target\n\n:Attribute Information:\n    - MedInc        median income in block group\n    - HouseAge      median house age in block group\n    - AveRooms      average number of rooms per household\n    - AveBedrms     average number of bedrooms per household\n    - Population    block group population\n    - AveOccup      average number of household members\n    - Latitude      block group latitude\n    - Longitude     block group longitude\n\n:Missing Attribute Values: None\n\nThis dataset was obtained from the StatLib repository.\nhttps://www.dcc.fc.up.pt/~ltorgo/Regression/cal_housing.html\n\nThe target variable is the median house value for California districts,\nexpressed in hundreds of thousands of dollars ($100,000).\n\nThis dataset was derived from the 1990 U.S

In [ ]:
housing_data.data[:4]

array([[ 8.32520000e+00,  4.10000000e+01,  6.98412698e+00,
         1.02380952e+00,  3.22000000e+02,  2.55555556e+00,
         3.78800000e+01, -1.22230000e+02],
       [ 8.30140000e+00,  2.10000000e+01,  6.23813708e+00,
         9.71880492e-01,  2.40100000e+03,  2.10984183e+00,
         3.78600000e+01, -1.22220000e+02],
       [ 7.25740000e+00,  5.20000000e+01,  8.28813559e+00,
         1.07344633e+00,  4.96000000e+02,  2.80225989e+00,
         3.78500000e+01, -1.22240000e+02],
       [ 5.64310000e+00,  5.20000000e+01,  5.81735160e+00,
         1.07305936e+00,  5.58000000e+02,  2.54794521e+00,
         3.78500000e+01, -1.22250000e+02]])

**Findout the shape of the dataset**

In [ ]:
housing_data.target.shape

(20640,)

In [ ]:
housing_data.target[:4]

array([4.526, 3.585, 3.521, 3.413])

In [ ]:
housing_data.feature_names

['MedInc',
 'HouseAge',
 'AveRooms',
 'AveBedrms',
 'Population',
 'AveOccup',
 'Latitude',
 'Longitude']

In [ ]:
housing_data.target_names

['MedHouseVal']

In [ ]:
import pandas as pd

pd.io

<module 'pandas.io' from '/usr/local/lib/python3.12/dist-packages/pandas/io/__init__.py'>

**fetch_openml**
1. [openml.org](https://) is public repositry for machine learing data and experiment, that allows everybody to upload open dataset
2. import the library and access the documentation.

In [1]:
from sklearn.datasets import fetch_openml
?fetch_openml

**Note that this is an experimental API and is likely to change in the future releases**
1. we use this API for loading MNIST dataset.

In [2]:
X, y = fetch_openml('mnist_784',version=1, return_X_y=True)
print('feature matrix shape:', X.shape)
print('targert matrix shape:', y.shape)

feature matrix shape: (70000, 784)
targert matrix shape: (70000,)


**Exercise**

fetch_20newsgroups

In [8]:
from sklearn.datasets import fetch_20newsgroups

data = fetch_20newsgroups()

X = data.data      # text data (news articles)
y = data.target    # labels (categories)

print(X[:2])
print(y[:5])

["From: lerxst@wam.umd.edu (where's my thing)\nSubject: WHAT car is this!?\nNntp-Posting-Host: rac3.wam.umd.edu\nOrganization: University of Maryland, College Park\nLines: 15\n\n I was wondering if anyone out there could enlighten me on this car I saw\nthe other day. It was a 2-door sports car, looked to be from the late 60s/\nearly 70s. It was called a Bricklin. The doors were really small. In addition,\nthe front bumper was separate from the rest of the body. This is \nall I know. If anyone can tellme a model name, engine specs, years\nof production, where this car is made, history, or whatever info you\nhave on this funky looking car, please e-mail.\n\nThanks,\n- IL\n   ---- brought to you by your neighborhood Lerxst ----\n\n\n\n\n", "From: guykuo@carson.u.washington.edu (Guy Kuo)\nSubject: SI Clock Poll - Final Call\nSummary: Final call for SI clock reports\nKeywords: SI,acceleration,clock,upgrade\nArticle-I.D.: shelley.1qvfo9INNc3s\nOrganization: University of Washington\nLines: 1

**Generators**
1. make_regression

In [9]:
from sklearn.datasets import make_regression
?make_regression

**Example 1**
1. Let's generate 100 samples with 5 features for a single label regression problem

In [12]:
X, y = make_regression(n_samples=100,n_features=5,n_targets=1,shuffle=True, random_state=42)

**its a good practice to set seed so that we get to see repeatability in the experimentation.**
1. Let's look at the shape of feature  matrix and label_vector

In [13]:
X.shape

(100, 5)

In [14]:
y.shape

(100,)

let's generate 100 sample with 5 features for multiple regression problem with 5 outputs

In [16]:
X,y = make_regression(n_samples=100,n_features=5,n_targets=4,shuffle=True,random_state=42)

In [17]:
X.shape

(100, 5)

In [18]:
y.shape

(100, 4)

**since we generate multi output target with 5 outputs, the out has shape (100, 4)**

**make_classification**
1. Generate a random n-class classification problem set up

In [19]:
from sklearn.datasets import make_classification
?make_classification

**Let's generate a binary classification problem with 10 features and 100 samples**

In [20]:
X, y = make_classification(n_samples=100, n_features=10, n_classes=2, n_clusters_per_class=1, random_state=42)

In [21]:
X.shape

(100, 10)

In [22]:
y.shape

(100,)

look at a few example and their labels

In [23]:
X[:5]

array([[ 0.11422765, -1.71016839, -0.06822216, -0.14928517,  0.30780177,
         0.15030176, -0.05694562, -0.22595246, -0.36361221, -0.13818757],
       [ 0.70775194, -1.57022472, -0.23503183, -0.63604713,  0.62180996,
        -0.56246678,  0.97255445, -0.77719676,  0.63240774, -0.47809669],
       [ 0.63859246,  0.04739867,  0.33273433,  1.1046981 , -0.65183611,
        -1.66152006, -1.2110162 ,  1.09821151, -0.0660798 ,  0.68024225],
       [-0.23894805, -0.97755524,  0.0379061 ,  0.19896733,  0.50091719,
        -0.90756366,  0.75539123,  0.12437227, -0.57677133,  0.07871283],
       [-0.59239392, -0.05023811,  0.17573204, -1.43949185,  0.27045683,
        -0.86399077, -0.83095012,  0.60046915,  0.04852163,  0.32557953]])

In [24]:
y[:5]

array([1, 1, 1, 1, 0])

**let's create 3 classes**

In [26]:
X, y = make_classification(n_samples=100, n_features=10, n_classes=3, n_clusters_per_class=1,random_state=42)

In [27]:
X.shape

(100, 10)

In [28]:
y.shape

(100,)

look at few ex. feature and label

In [29]:
X[:5]

array([[-0.58351628, -1.73833907, -1.37298251, -1.77311485,  0.45918008,
         0.83392215, -1.66096093,  0.20768769, -0.07016571,  0.42961822],
       [-1.0044394 , -1.43862044,  0.47335819, -0.21188291,  0.0125924 ,
         0.22409248, -0.77300978,  0.49799829,  0.0976761 ,  0.02451017],
       [ 0.07740833,  0.19896733,  0.12437227,  0.17738132, -0.97755524,
         0.50091719,  0.75138712,  0.54336019,  0.09933231, -1.66940528],
       [-0.91759569, -0.9609536 ,  1.07746664,  0.4522739 , -0.32138584,
        -0.8254972 , -0.56372455,  0.24368721,  0.41293145, -0.8222204 ],
       [-0.96222828, -0.96090774,  1.21530116,  0.55980482, -1.24778318,
        -0.25256815, -1.43014138,  0.13074058,  1.6324113 , -0.44004449]])

In [30]:
y[:5]

array([2, 0, 1, 0, 0])

**make_multilabel_classification**
1. This function helps us generate a random multi-label clssfication problem.

In [31]:
from sklearn.datasets import make_multilabel_classification
?make_multilabel_classification

Let's generate a multilabel classification problem with 100 samples, 10 features, 5 labels and on an average 2 labels per exaple.

In [32]:
X, y = make_multilabel_classification(n_samples=100, n_features=20, n_classes=5, n_labels=2)

**First of all let's examine shape of feature matrix and label vector.**

In [33]:
X.shape

(100, 20)

In [34]:
y.shape

(100, 5)

In [35]:
X[:5]

array([[1., 4., 2., 2., 0., 2., 0., 7., 7., 1., 2., 4., 2., 3., 2., 6.,
        0., 3., 1., 3.],
       [2., 4., 3., 0., 0., 3., 1., 5., 7., 2., 5., 2., 1., 1., 2., 2.,
        1., 1., 1., 5.],
       [5., 3., 1., 1., 1., 3., 1., 3., 2., 4., 3., 2., 2., 5., 3., 7.,
        3., 1., 2., 0.],
       [1., 2., 2., 2., 0., 0., 0., 4., 4., 0., 1., 1., 3., 1., 6., 6.,
        3., 3., 2., 1.],
       [5., 3., 1., 1., 0., 0., 2., 3., 4., 3., 3., 0., 3., 7., 3., 1.,
        1., 1., 2., 7.]])

In [36]:
y[:5]

array([[1, 0, 0, 0, 0],
       [1, 0, 0, 0, 0],
       [1, 0, 0, 0, 0],
       [1, 0, 0, 1, 0],
       [0, 1, 0, 0, 0]])

**make_blobs**
1. make_blobs enables us to generate random data for clustering

In [37]:
from sklearn.datasets import make_blobs
?make_blobs

Let's generate a random dataset of 10 samples with 2 features each for clustering

In [39]:
X, y = make_blobs(n_samples=10, centers=3, n_features=2, random_state=42)
print('feature matrix shape:', X.shape)
print('label shaoe:', y.shape)

feature matrix shape: (10, 2)
label shaoe: (10,)


we can find the cluster memrship of each point in y

In [41]:
y

array([2, 2, 1, 2, 0, 0, 0, 1, 1, 0])